pkg/controller/seed-controller-manager/etcdrestore/etcd_restore_controller.go

```diff
@@ -0,0 +1,267 @@
+	workerName string
+	ctrlruntimeclient.Client
+	recorder record.EventRecorder
+}
+
+// Add creates a new etcd restore controller that is responsible for
+// managing cluster etcd restores
+func Add(
+	mgr manager.Manager,
+	log *zap.SugaredLogger,
+	numWorkers int,
+	workerName string,
+	s3Endpoint string,
+	s3BucketName string,
+	s3AccessKeyID string,
+	s3SecretAccessKey string,
+) error {
+	log = log.Named(ControllerName)
+	client := mgr.GetClient()
+
+	reconciler := &Reconciler{
+		log:        log,
+		Client:     client,
+		workerName: workerName,
+		recorder:   mgr.GetEventRecorderFor(ControllerName),
+	}
+
+	ctrlOptions := controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	}
+	c, err := controller.New(ControllerName, mgr, ctrlOptions)
+	if err != nil {
+		return err
+	}
+
+	incompleteRestorePredicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			restore := e.Object.(*kubermaticv1.EtcdRestore)
+			return restore.Status.Phase != kubermaticv1.EtcdRestorePhaseCompleted
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			restore := e.ObjectNew.(*kubermaticv1.EtcdRestore)
+			return restore.Status.Phase != kubermaticv1.EtcdRestorePhaseCompleted
+		},
+	}
+
+	return c.Watch(&source.Kind{Type: &kubermaticv1.EtcdRestore{}}, &handler.EnqueueRequestForObject{}, incompleteRestorePredicates)
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	restore := &kubermaticv1.EtcdRestore{}
+	if err := r.Get(ctx, request.NamespacedName, restore); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, types.NamespacedName{Name: restore.Spec.Cluster.Name}, cluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	log = r.log.With("cluster", cluster.Name, "restore", restore.Name)
+
+	if cluster.Labels[kubermaticv1.WorkerNameLabelKey] != r.workerName {
+		return reconcile.Result{}, nil
+	}
+
+	result, err := r.reconcile(ctx, log, restore, cluster)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.recorder.Event(restore, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+		r.recorder.Eventf(cluster, corev1.EventTypeWarning, "ReconcilingError",
+			"failed to reconcile etcd restore %q: %v", restore.Name, err)
+	}
+
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, restore *kubermaticv1.EtcdRestore, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	log.Infof("performing etcd restore from backup %v", restore.Spec.BackupName)
+
+	if restore.Status.Phase == kubermaticv1.EtcdRestorePhaseCompleted {
+		return nil, nil
+	}
+
```

Comment:
I think we should try to make sure we have everything we need before starting the destructive actions. We should at least check if we have the settings and the secrets available.

Reply:
The controller now checks that the backup exists in S3 before proceeding here.